File: 108. Convert Sorted Array to Binary Search Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	root := helper(nums, 0, len(nums)-1)
	return root
}

func helper(nums []int, start, end int) *TreeNode {
	if start > end {
		return nil
	}
	mid := (start + end) / 2
	val := nums[mid]
	node := &TreeNode{Val: val, Left: nil, Right: nil}
	node.Left = helper(nums, start, mid-1)
	node.Right = helper(nums, mid+1, end)
	return node
}

File: 108. Convert Sorted Array to Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def helper(self, num, first, last):
        if last < first:
            return None

        mid = first + (last - first) // 2
        root = TreeNode(num[mid])
        root.left = self.helper(num, first, mid - 1)
        root.right = self.helper(num, mid + 1, last)
        return root

    def sortedArrayToBST(self, num):
        return self.helper(num, 0, len(num) - 1)
